### PR TITLE
IoP blockchain parameters and premining

### PR DESCRIPTION
**Pull request by acostarodrigo**
# Blockchain parameters
* Added specific IoP blockchain parameters for MainNet, TestNet and RegTest networks.
* Changed default port numbers
* changed Genesis block timestamp and script.
* changed magic bytes, checkpoints and base58 prefixes.
* changed default seed nodes.

# Added premine functionality at first generated block.
* Changed functionality to premine at block 1.

### CODE REVIEW

**Comment by droka** [comment]
Didnt you want to change the nRPCPort from 8337 to 4337 ?

**Comment by droka** [comment]
There were different additions and changes to the original bitcoin protocol, that came into effect at a  later block. These seem to be set up in the chainparams file to turn on at a certain block number.
Those must be changed in our version to either turn on at block zero, or not to turn on at all. Otherwise the iop blockchain will only turn these features on at some later point, like bitcoin did.

**Comment by droka** [comment]
I would rather put the premine into genesis block. Otherwise every node, ever processing any block in the future, have to do an additional "if".

**Comment by izolyomi** [comment]
Won't it be better to do the premining in the genesis block and just change the genesis block to be spendable? Probably this would express our intentions more directly and would need no special checks in the code. For how to make the genesis block spendable, see e.g. http://bitcoin.stackexchange.com/questions/37559/how-to-make-genesis-block-transaction-spendable

File: src/main.cpp [hunk 1]
```diff
@@ -2543,6 +2631,65 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
+		const std::vector<unsigned char> signature = value;
+
+		if (!IsValidSignatureEncoding(value)){
+			LogPrint("Invalid coinbase transaction", "Provided signature is not valid: %s \n", HexStr(value));
+			return state.DoS(100, false, REJECT_INVALID, "bad-CB-signature", false, "Coinbase invalid signature");
+		}
+
+		//we remove the sig hash type from the end of the signature
+		vector<unsigned char> vchSig(value);
+		vchSig.back();
```
Comment (hellwolf) [inline src/main.cpp]: This line seems useless.

File: src/main.cpp [hunk 2]
```diff
@@ -2531,6 +2551,74 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
...
+							while (pc < out.scriptPubKey.end()){
+								out.scriptPubKey.GetOp(pc, opcode, value);
+							}
+
+							// we get the OP_Return data into the string. Maybe this can be improved
+							std::string opreturn = StrHex(HexStr(value));
+							LogPrint("MinerWhiteListTransaction", "OP_RETURN data: %s \n", opreturn);
+							if (opreturn.size() > 3){
+								CMinerWhiteList minerwhitelistdb;
+								minerwhitelist_v vector;
```
Comment (hellwolf) [inline src/main.cpp]: nit picking here... would try to avoid using "vector", which can be confusing  with std::vector used around.

**Comment by acostarodrigo** [comment]
Regarding your comments about including the premining on the genesis block instead if block 1, I agree. And originally I tried to do it this way. I imported the private key that was used to generate the genesis block and got the premine on my wallet. 

But the genesis block is not stored on the blockchain or Berkeley DB , it is hardcoded inside. 

So no client accepted a transaction using tokens from the genesis block because it was not found at the blockchain. A change in how validations are made needs to be introduced. This is probably better than doing the if statement for every block into the future, but I didn't have time to look at it. If anyone can take a look an propose a change, even better.

**Comment by droka** [comment]
Namecoin also choose not to rename the "bitcoin" words in the code.
https://github.com/namecoin/namecoin-legacy/search?p=1&q=bitcoin&utf8=%E2%9C%93

**Comment by izolyomi** [comment]
Just tried making the genesis block spendable. Rodrigo was right, it's not as easy as stated [here](http://bitcoin.stackexchange.com/questions/37559/how-to-make-genesis-block-transaction-spendable), segfaults on startup even with an empty block db. Let's keep premining the foundation millions in the first block.

**Pull request merged**